Jumpstart ag - deployment readme (#1916)
microsoft/azure_arc: docs/azure_jumpstart_ag/contoso_supermarket/deployment/_index.md

toc_hide: true
---

## Jumpstart Agora - Deployment guide

> Jumpstart Agora provides a simple deployment process using Azure Bicep and PowerShell that minimizes user interaction. This automation automatically configures the Agora environment, including the infrastructure, the Contoso-Supermarket applications, CI/CD artifacts, and observability components. The diagram below details the high-level architecture that is deployed and configured as part of the automation.

Reviewer: ```suggestion
Jumpstart Agora provides a simple deployment process using Azure Bicep and PowerShell that minimizes user interaction. This automation automatically configures the Contoso Supermarket scenario environment, including the infrastructure, the Contoso-Supermarket applications, CI/CD artifacts, and observability components. The diagram below details the high-level architecture that is deployed and configured as part of the automation.
```

![Architecture diagram](./img/architecture_diagram.png)

Deploying Agora consists of the following steps.

>     1. Prerequisite - User creates or already has access to an Azure service principal with Owner role scoped to an Azure subscription.
>     2. Prerequisite - User forks the [Jumpstart-Agora-Apps repo](https://github.com/microsoft/jumpstart-agora-apps) into their own GitHub account.
>     3. Deploy infrastructure - User deploys a Bicep file that creates the infrastructure in an Azure resource group.
>     4. Login to the Agora-VM-Client Azure virtual machine - On first logon to the Agora-VM-Client PowerShell scripts automatically run that configure the applications and CI/CD. These scripts will take some time to run.
>     5. Automation completes - The PowerShell windows automatically close and the Agora desktop background is visible.

Reviewer: ```suggestion
    1. Prerequisite - User creates or already has access to an Azure service principal with Owner role scoped to an Azure subscription.
    2. Prerequisite - User forks the [Jumpstart-Agora-Apps GitHub repository](https://github.com/microsoft/jumpstart-agora-apps) into their own GitHub account.
    3. Deploy infrastructure - User deploys a Bicep template that creates the infrastructure in an Azure resource group.
    4. Login to the Agora-VM-Client Azure virtual machine - On the first logon to the Agora-VM-Client PowerShell scripts automatically run that configure the applications and CI/CD. These scripts will take some time to run.
    5. Automation completes - The PowerShell windows automatically close and the Agora desktop background is visible.
```

Once automation is complete, users can immediately start enjoying the Contoso Supermarket experience.

![Architecture diagram](./img/deployment_flow.png)

## Prerequisites

> - Fork the [Jumpstart-Agora-Apps repo](https://github.com/microsoft/jumpstart-agora-apps) into your own GitHub account. If you do not have a GitHub account, visit [this link](https://github.com/join) to create one.

Reviewer: ```suggestion
- Fork the [_Jumpstart-Agora-Apps_ repo](https://github.com/microsoft/jumpstart-agora-apps) into your own GitHub account. If you do not have a GitHub account, visit [this link](https://github.com/join) to create one.
```

>   - Open the [Jumpstart-Agora-Apps repo](https://github.com/microsoft/jumpstart-agora-apps) while signed into your GitHub account and click the "Fork" button in the upper-right corner.

Reviewer: ```suggestion
  - Open the [_Jumpstart-Agora-Apps_ repo](https://github.com/microsoft/jumpstart-agora-apps) while signed into your GitHub account and click the "Fork" button in the upper-right corner.
```

    ![Screenshot showing how to fork the Jumpstart Agora Apps repo](./img/fork_repo1.png)
  
  - Use the default settings to create your fork. Ensure that your user is selected in the dropdown and then click the "Create Fork" button.

    ![Screenshot showing how to fork the Jumpstart Agora Apps repo](./img/fork_repo2.png)

> - Configure a GitHub fine-grained personal access token (PAT) with permissions to modify *only* the Jumpstart Agora Apps repo that you forked.

Reviewer: I would put a note here around being in beta

  - In the top right of the GitHub website, click on the dropdown on your user icon and then click "Settings".

    ![Screenshot showing how to create the GitHub PAT](./img/github_PAT1.png)

  - In the left menu navigation click on "Developer settings".

    ![Screenshot showing how to create the GitHub PAT](./img/github_PAT2.png)

  - Under "Personal access tokens" select "Fine-grained tokens".

    ![Screenshot showing how to create the GitHub PAT](./img/github_PAT3.png)

    ![Screenshot showing how to create the GitHub PAT](./img/github_PAT4.png)

  - Give your token a name and select an expiry period.

    ![Screenshot showing how to create the GitHub PAT](./img/github_PAT5.png)

>   - Under "Repository access" select "Only select repositories" and then choose the jumpstart-agora-apps repo that you forked in the previous steps.

Reviewer: ```suggestion
  - Under "Repository access" select "Only select repositories" and then choose the _jumpstart-agora-apps_ repo that you forked in the previous steps.
```

    ![Screenshot showing how to create the GitHub PAT](./img/github_PAT6.png)

>   - Assign "Read and write" permissions for your fork of the jumpstart-agora-apps repo. The PAT should have read and write permissions for the following operations: Actions, Administration, Contents, Pull Requests, Secrets, Workflows.

Reviewer: ```suggestion
  - Assign the "Read and write" permissions for your fork of the _jumpstart-agora-apps_ repo. The PAT should have read and write permissions for the following operations: Actions, Administration, Contents, Pull Requests, Secrets, and Workflows.
```

  ```

## Deployment: Bicep deployment via Azure CLI

- Clone the Azure Arc Jumpstart repository

  ```shell
  git clone https://github.com/microsoft/azure_arc.git
  ```

- Upgrade to latest Bicep version

  ```shell
  az bicep upgrade
  ```

- Edit the [main.parameters.json](https://github.com/microsoft/azure_arc/blob/jumpstart_ag/azure_jumpstart_ag/bicep/main.parameters.json) template parameters file and supply some values for your environment.
  - _`sshRSAPublicKey`_ - Your SSH public key
  - _`spnClientId`_ - Your Azure service principal id
  - _`spnClientSecret`_ - Your Azure service principal secret
  - _`spnTenantId`_ - Your Azure tenant id
  - _`windowsAdminUsername`_ - Client Windows VM Administrator name
  - _`windowsAdminPassword`_ - Client Windows VM Password. Password must have 3 of the following: 1 lower case character, 1 upper case character, 1 number, and 1 special character. The value must be between 12 and 123 characters long.
  - _`deployBastion`_ - Option to deploy using Azure Bastion instead of traditional RDP. Set to true or false.

>   - _`githubUser`_ - The name of the user account on GitHub where you have forked the [Jumpstart-Agora-Apps repo](https://github.com/microsoft/jumpstart-agora-apps)

Reviewer: ```suggestion
  - _`githubUser`_ - The name of the user account on GitHub where you have forked the [_Jumpstart-Agora-Apps_ repo](https://github.com/microsoft/jumpstart-agora-apps)
```

  - _`githubPAT`_ - The GitHub PAT token that you created as part of the prerequisites.

  ![Screenshot showing example parameters](./img/parameters_bicep.png)

> - Now you will deploy the Bicep file. Navigate to the local cloned [deployment folder](https://github.com/microsoft/azure_arc/tree/main/azure_jumpstart_ag/bicep) and run the below command:

Reviewer: We should change the path until we will merge to main

  ```shell
  az login
  az group create --name "<resource-group-name>"  --location "<preferred-location>"
  az deployment group create -g "<resource-group-name>" -f "main.bicep" -p "main.parameters.json"
  ```

    > **NOTE: If you see any failure in the deployment, please check the [troubleshooting guide](https://placeholder).**

## Start post-deployment automation

> Once your deployment is complete, you can open the Azure portal and see the Agora resources inside your resource group. You will be using the _Agora-Client-VM_ Azure virtual machine to explore various capabilities of Agora. You will need to remotely access _Agora-Client-VM_.

Reviewer: ```suggestion
Once your deployment is complete, you can open the Azure portal and see the Agora scenario resources inside your resource group. You will be using the _Agora-Client-VM_ Azure virtual machine to explore various capabilities of Agora. You will need to remotely access _Agora-Client-VM_.
```
